parsers/v.c

```diff
@@ -0,0 +1,3011 @@
+			if (c == '>')
+				getcAndCaptureIfEq ('>', capture, token);
+			if (getcAndCaptureIfEq ('=', capture, token))
+				token->type = TOKEN_ASSIGN; // <<= >>= >>>=
+			else
+				token->type = TOKEN_OPERATOR; // << >> >>>
+		}
+		else if (c == '<' && getcAndCaptureIfEq ('-', capture, token))
+			token->type = TOKEN_CHPOP; // <-
+		else
+		{
+			getcAndCaptureIfEq ('=', capture, token);
+			token->type = TOKEN_OPERATOR; // < > <= >=
+		}
+	}
+	else if (c == '.')
+	{
+		if (getcAndCaptureIfEq ('.', capture, token))
+		{
+			getcAndCaptureIfEq ('.', capture, token);
+			token->type = TOKEN_SLICE; // .. ...
+		}
+		else
+			token->type = TOKEN_DOT; // .
+	}
+	else if (c == ':')
+	{
+		if (getcAndCaptureIfEq ('=', capture, token))
+			token->type = TOKEN_DECLARE; // :=
+		else
+			token->type = TOKEN_COLON; // :
+	}
+	else if (isOneOf (c, "\"'`"))
+	{
+		int terminator = c;
+		bool inEscape = false;
+		bool inEval = false;
+		int evalLevel = 0;
+		token->type = TOKEN_IMMEDIATE;
+		do
+		{
+			if (inEscape)
+			{
+				captureChar (capture, token, c);
+				inEscape = false;
+			}
+			else if (inEval) // TODO: this doesn't handle nested strings/evals
+			{
+				if (c == '{')
+					evalLevel++;
+				else if (c == '}')
+					evalLevel--;
+				if (evalLevel == 0)
+					inEval = false;
+			}
+			else if (c == '\\')
+				inEscape = true;
+			else if (c == '$')
+				inEval = true;
+			c = getcFromInputFile ();
+			captureChar (capture, token, c);
+		} while (c != EOF &&
+				 (c != terminator || inEscape || (inEval && evalLevel > 0)));
+		checkLabel (token, capture);
+	}
+	else if (c == '@')
+	{
+		if (isSubsequentIdent (peekcFromInputFile ()))
+		{
+			token->type = TOKEN_IDENT;
+			token->string = vStringNew ();
+			bool more;
+			do
+			{
+				vStringPut (token->string, c);
+				c = getcFromInputFile ();
+				more = isSubsequentIdent (c);
+				if (more)
+					captureChar (capture, token, c);
+			} while (more);
+			ungetcToInputFile (c);
+			checkLabel (token, capture);
+		}
+		else
+			token->type = TOKEN_AT;
+	}
+	else if (isInitialIdent (c))
+	{
+		token->string = vStringNew ();
+		bool more;
+		do
+		{
+			vStringPut (token->string, c);
+			c = getcFromInputFile ();
+			more = isSubsequentIdent (c);
+			if (more)
+				captureChar (capture, token, c);
+		} while (more);
+		ungetcToInputFile (c);
+
+		token->keyword = lookupKeyword (vStringValue (token->string), Lang_v);
+		if (token->keyword != KEYWORD_NONE)
+			token->type = TOKEN_KEYWORD;
+		else
+		{
+			char initChar = vStringChar (token->string, 0);
+			if (initChar >= 'A' && initChar <= 'Z')
+				token->type = TOKEN_TYPE;
+			else if (isOneOf (initChar, "rc") &&
+					 vStringLength (token->string) == 1 &&
+					 isOneOf (peekcFromInputFile (), "'\"`"))
+			{
+				int end = getcFromInputFile ();
+				captureChar (capture, token, end);
+				do
+				{
+					c = getcFromInputFile ();
+					captureChar (capture, token, c);
+				}
+				while (c != end && c != EOF);
+				vStringDelete (token->string);
```

Comment:
Of course, this is correct C code. However, I don't like this. Allocating and Deleting token->string doesn't follow the Ctags' code pattern.

In many MODERN parsers, the life cycles of `token` and `token->string` are the same.
This is not a must. However, this makes the life of people (who?) maintaining multiple parsers easier. 

"Keeping a token as-is but replacing its `vstring`" looks tricky to me.

Please compare clearPoolToken() of v.c and clearPoolToken() in python.c.
v.c uses vStringDelete. However, pythons.c v.c uses vStringClear.

Reply:
The V parser does not have a string for many tokens, so I defaulted it to NULL. Happy to change though.